src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -347,149 +349,147 @@ private Socket TryConnectParallel(string hostName, int port, TimeSpan ts, bool i
+                    Blocking = isInfiniteTimeout
+                };
 
-            Socket[] sockets = new Socket[ipAddresses.Length];
-            AddressFamily[] preferedIPFamilies = new AddressFamily[2];
+                bool isSocketSelected = false;
 
-            if (ipPreference == SqlConnectionIPAddressPreference.IPv4First)
-            {
-                preferedIPFamilies[0] = AddressFamily.InterNetwork;
-                preferedIPFamilies[1] = AddressFamily.InterNetworkV6;
-            }
-            else if (ipPreference == SqlConnectionIPAddressPreference.IPv6First)
-            {
-                preferedIPFamilies[0] = AddressFamily.InterNetworkV6;
-                preferedIPFamilies[1] = AddressFamily.InterNetwork;
-            }
-            // else -> UsePlatformDefault
+                // enable keep-alive on socket
+                SetKeepAliveValues(ref socket);
 
-            CancellationTokenSource cts = null;
+                SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNITCPHandle), EventType.INFO,
+                    "Connecting to IP address {0} and port {1} using {2} address family. Is infinite timeout: {3}",
+                    ipAddress,
+                    port, 
+                    ipAddress.AddressFamily, 
+                    isInfiniteTimeout);
 
-            if (!isInfiniteTimeout)
-            {
-                cts = new CancellationTokenSource(timeout);
-                cts.Token.Register(Cancel);
-            }
 
-            Socket availableSocket = null;
-            try
-            {
-                // We go through the IP list twice.
-                // In the first traversal, we only try to connect with the preferedIPFamilies[0].
-                // In the second traversal, we only try to connect with the preferedIPFamilies[1].
-                // For UsePlatformDefault preference, we do traversal once.
-                for (int i = 0; i < preferedIPFamilies.Length; ++i)
+                try
                 {
-                    for (int n = 0; n < ipAddresses.Length; n++)
+                    bool isConnected;
+                    try // catching SocketException with SocketErrorCode == WouldBlock to run Socket.Select
                     {
-                        IPAddress ipAddress = ipAddresses[n];
-                        try
+                        socket.Connect(ipAddress, port);
+                        if (!isInfiniteTimeout)
+                            throw SQL.SocketDidNotThrow();
+                        
+                        isConnected = true;
+                    }
+                    catch (SocketException socketException) when (!isInfiniteTimeout &&
+                                                                  socketException.SocketErrorCode ==
+                                                                  SocketError.WouldBlock)
+                    {
+                        // https://github.com/dotnet/SqlClient/issues/826#issuecomment-736224118
```

Comment:
Please, include the outcome of the discussions as the comments in the code. For reference, I'd put links as review notes for reviewers.